test/e2e/user_certs_test.go

```diff
@@ -0,0 +1,281 @@
+		"three":        servers["three"].Certificate.SerialNumber.String(),
+		"serving-cert": serialNumberOfCertificateFromSecretOrFail(t, kubeClient, "openshift-kube-apiserver", "serving-cert"),
+	}
+
+	// execute test cases
+	testCases := []struct {
+		name                 string
+		serverName           string
+		expectedSerialNumber string
+	}{
+		{
+			name:                 "User one.test",
+			serverName:           "one.test",
+			expectedSerialNumber: serialNumbers["one"],
+		},
+		{
+			name:                 "User two.test",
+			serverName:           "two.test",
+			expectedSerialNumber: serialNumbers["two"],
+		},
+		{
+			name:                 "User three.test",
+			serverName:           "three.test",
+			expectedSerialNumber: serialNumbers["three"],
+		},
+		{
+			name:                 "User four.test",
+			serverName:           "four.test",
+			expectedSerialNumber: serialNumbers["three"],
+		},
+		{
+			name:                 "Service kubernetes",
+			serverName:           "kubernetes",
+			expectedSerialNumber: serialNumbers["serving-cert"],
+		},
+		{
+			name:                 "Service kubernetes.default",
+			serverName:           "kubernetes.default",
+			expectedSerialNumber: serialNumbers["serving-cert"],
+		},
+		{
+			name:                 "Service kubernetes.default.svc",
+			serverName:           "kubernetes.default.svc",
+			expectedSerialNumber: serialNumbers["serving-cert"],
+		},
+		{
+			name:                 "Service kubernetes.default.svc.cluster.local",
+			serverName:           "kubernetes.default.svc.cluster.local",
+			expectedSerialNumber: serialNumbers["serving-cert"],
+		},
+		{
+			name:                 "ServiceIP",
+			serverName:           getKubernetesServiceClusterIPOrFail(t, kubeClient),
+			expectedSerialNumber: serialNumbers["serving-cert"],
+		},
+		{
+			name:                 "Localhost localhost",
+			serverName:           "localhost",
+			expectedSerialNumber: serialNumbers["serving-cert"],
+		},
+		{
+			name:                 "Localhost 127.0.0.1",
+			serverName:           "127.0.0.1",
+			expectedSerialNumber: serialNumbers["serving-cert"],
+		},
+		{
+			name:                 "LoadBalancerHostname",
+			serverName:           getAPIServiceHostNameOrFail(t, configClient),
+			expectedSerialNumber: serialNumbers["serving-cert"],
+		},
+		{
+			name:                 "UnknownServerHostname",
+			serverName:           "unknown.test",
+			expectedSerialNumber: serialNumbers["serving-cert"],
+		},
+	}
+
+	// execute test cases
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// keep track of the incoming certificate's serial number
+			var serialNumber string
+			verifyPeerCertificate := func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+				var err error
+				if certificate, err := x509.ParseCertificate(rawCerts[0]); err == nil {
+					serialNumber = certificate.SerialNumber.String()
+				}
+				return err
+			}
+			tlsConf := &tls.Config{
```

Comment:
// connect to server with custom ServerName and check serial number of used server cert